RetryWhen with Retry.fixedDelay may fail with Spec Rule 1.3

This is with latest and greatest with boot 2.4.0.

In a Spring Statemachine I'm using `EmitterProcessor` via Sinks api as `Sinks.many().multicast().onBackpressureBuffer(Queues.SMALL_BUFFER_SIZE, false)`.

Javadocs mentions that I could sleep to go back to original behaviour and then I don't have an issue:
```
while (triggerSink.tryEmitNext(tqi).isFailure()) {
  LockSupport.parkNanos(10);
}
```

However this naturally gets `blockhound` to complain about blocking operation so I've been trying to find a non-blocking alternative. My idea was an attempt to try emit and then retry few times with a delay. If it magically goes through and doesn't error with serial access, good and otherwise I'd propagate with some error handling fu.

```
.flatMap(tqi -> Mono.fromRunnable(() -> {
    triggerSink.emitNext(tqi, EmitFailureHandler.FAIL_FAST);
  })
  .retryWhen(Retry.fixedDelay(10, Duration.ofNanos(10)))
  )
```

However it looks like when trying to use `retryWhen` with `Retry.fixedDelay`, reactor internally may get into same problem I'm just trying to overcome with retrying emit operation. I was able to come up a test outside of a statemachine to show this same behaviour and pushed project to https://github.com/jvalkeal/randomstuff/tree/master/reactor-retrywhen-issue:

Running a test there doesn't always fail and may actually return proper error(`RetryExhaustedException`) but when you run it many times, you should be able to see it fail with:


```
org.opentest4j.AssertionFailedError: 
Expecting:
 <java.lang.AssertionError: VerifySubscriber timed out on reactor.core.publisher.MonoIgnoreElements$IgnoreElementsSubscriber@1cd9f0f1>
to be equal to:
 <null>
but was not.
```

Then you'd be able to find this spesific error logged by reactor itself:

```
13:40:07.880 [parallel-2] ERROR reactor.core.publisher.Operators - Operator called default onErrorDropped
reactor.core.publisher.Sinks$EmissionException: Spec. Rule 1.3 - onSubscribe, onNext, onError and onComplete signaled to a Subscriber MUST be signaled serially.
	at reactor.core.publisher.InternalManySink.emitNext(InternalManySink.java:56)
	at reactor.core.publisher.FluxRetryWhen$RetryWhenMainSubscriber.onError(FluxRetryWhen.java:191)
	at reactor.core.publisher.FluxPeek$PeekSubscriber.onError(FluxPeek.java:221)
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.checkTerminated(FluxFlatMap.java:841)
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.drainLoop(FluxFlatMap.java:607)
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.drain(FluxFlatMap.java:587)
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.onError(FluxFlatMap.java:450)
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.onNext(FluxFlatMap.java:415)
	at reactor.core.publisher.FluxPeekFuseable$PeekFuseableSubscriber.onNext(FluxPeekFuseable.java:210)
	at reactor.core.publisher.FluxArray$ArraySubscription.slowPath(FluxArray.java:127)
	at reactor.core.publisher.FluxArray$ArraySubscription.request(FluxArray.java:100)
	at reactor.core.publisher.FluxPeekFuseable$PeekFuseableSubscriber.request(FluxPeekFuseable.java:144)
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.onSubscribe(FluxFlatMap.java:370)
	at reactor.core.publisher.FluxPeekFuseable$PeekFuseableSubscriber.onSubscribe(FluxPeekFuseable.java:178)
	at reactor.core.publisher.FluxArray.subscribe(FluxArray.java:53)
	at reactor.core.publisher.FluxArray.subscribe(FluxArray.java:59)
	at reactor.core.publisher.InternalFluxOperator.subscribe(InternalFluxOperator.java:62)
	at reactor.core.publisher.FluxRetryWhen$RetryWhenMainSubscriber.resubscribe(FluxRetryWhen.java:215)
	at reactor.core.publisher.FluxRetryWhen$RetryWhenOtherSubscriber.onNext(FluxRetryWhen.java:268)
	at reactor.core.publisher.FluxConcatMap$ConcatMapImmediate.innerNext(FluxConcatMap.java:281)
	at reactor.core.publisher.FluxConcatMap$ConcatMapInner.onNext(FluxConcatMap.java:860)
	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1784)
	at reactor.core.publisher.MonoFlatMap$FlatMapInner.onNext(MonoFlatMap.java:249)
	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1784)
	at reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.drain(MonoIgnoreThen.java:148)
	at reactor.core.publisher.MonoIgnoreThen.subscribe(MonoIgnoreThen.java:56)
	at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:157)
	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1784)
	at reactor.core.publisher.MonoIgnoreThen$ThenAcceptInner.onNext(MonoIgnoreThen.java:305)
	at reactor.core.publisher.MonoDelay$MonoDelayRunnable.run(MonoDelay.java:119)
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:68)
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:28)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

As `Retry.fixedDelay` fails internally with `RetryWhen` I can't find a way to handle it within user code anymore. If I understood this correctly, retry feature just stops.

--- a/reactor-core/src/main/java/reactor/core/publisher/FluxRetryWhen.java
+++ b/reactor-core/src/main/java/reactor/core/publisher/FluxRetryWhen.java
@@ -186,9 +186,9 @@ final class FluxRetryWhen<T> extends InternalFluxOperator<T, T> {
 				produced(p);
 			}
 
-			otherArbiter.request(1);
-
 			signaller.emitNext(this, Sinks.EmitFailureHandler.FAIL_FAST);
+			// request after signalling, otherwise it may race
+			otherArbiter.request(1);
 		}
 
 		@Override

--- a/reactor-core/src/test/java/reactor/core/publisher/FluxRetryWhenTest.java
+++ b/reactor-core/src/test/java/reactor/core/publisher/FluxRetryWhenTest.java
@@ -1130,7 +1130,7 @@ public class FluxRetryWhenTest {
 				    }
 			    })
 			    .retryWhen(Retry.fixedDelay(10, Duration.ofNanos(1)))
-			    .blockFirst(Duration.ofSeconds(2));
+			    .blockFirst(Duration.ofSeconds(1));
 		}
 	}
 

--- a/reactor-core/src/test/java/reactor/core/publisher/FluxRetryWhenTest.java
+++ b/reactor-core/src/test/java/reactor/core/publisher/FluxRetryWhenTest.java
@@ -1119,4 +1119,19 @@ public class FluxRetryWhenTest {
 		            .verifyComplete();
 	}
 
+	@Test
+	void gh2488() {
+		for (int i = 0; i < 1_000; i++) {
+			AtomicInteger sourceHelper = new AtomicInteger();
+			Flux.just("hello")
+			    .doOnNext(m -> {
+				    if (sourceHelper.getAndIncrement() < 9) {
+					    throw new RuntimeException("Boom!");
+				    }
+			    })
+			    .retryWhen(Retry.fixedDelay(10, Duration.ofNanos(1)))
+			    .blockFirst(Duration.ofSeconds(2));
+		}
+	}
+
 }

